src/fiemap.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2020 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::os::unix::io::RawFd;
+
+use serde::{Deserialize, Serialize};
+
+use crate::errors::*;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct Extent {
+    pub logical: u64,
+    pub physical: u64,
+    pub length: u64,
+}
+
+/// Returns the `Extent`s associated with the given file. Note that the physical offsets are
+/// relative to the partition start on which the file resides.
+pub fn fiemap(fd: RawFd) -> Result<Vec<Extent>> {
```

Comment:
Thanks! Weird I somehow missed that. Will take a look and compare approaches.

Reply:
OK, had a look at this. Some notes:

- Uses an iterator interface, which is neat. (One minor downside is that it hurts readability a bit because state maintenance is scattered across functions rather than just in one big loop).
- Uses fewer of the FFI facilities available. For example, it directly declares the `ioctl` function as an external function instead of the `nix::sys::ioctl` helpers and encodes the `FS_IOC_FIEMAP` constant instead of deriving it via the `nix` macros.

Overall, I don't find it obviously better (but TBF, not obviously worse either). We could probably upstream tweaks from here and rebase on top of that. Though not sure if I want to block on that.